### DISCUSSION TITLE
Cannot access all RAM in STM32 NUCLEO_H723ZG port

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#12597] [STM32 / Pyboard] [answered]
Hi,

I successfully built and flashed a NUCLEO_H723ZG board with MicroPython v1.20. 
When I check the RAM after a clean hard reset:
```python
import gc
gc.collect()
gc.mem_free()
```
I get: 74416 bytes

The [specs](https://www.st.com/en/microcontrollers-microprocessors/stm32h723-733.html) of the STM32H723ZG indicate:
- 28 Kbytes of data TCM RAM for critical real time data
- 432 Kbytes of system RAM (up to 256 Kbytes can remap on instruction TCM RAM for critical real time instructions)
- 4 Kbytes of backup SRAM (available in the lowest-power modes)

Any idea why I get no value near 400 kB?
Thanks!

**Comment by robert-hh** [comment]
The loader script seems to configur for 128k+128k RAM. So maybe that one has to be changed.

**Reply by saraverbeecke** [reply]
Thanks for pointing to that file @robert-hh.

stm32h723.ld contains:
```
...
/* Specify the memory areas */
MEMORY
{
    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 768K    /* sectors 0-5 */
    FLASH_APP  (rx) : ORIGIN = 0x08020000, LENGTH = 640K    /* sectors 1-5 */
    FLASH_FS (r)    : ORIGIN = 0x080c0000, LENGTH = 256K    /* sectors 6-7 */
    DTCM (xrw)      : ORIGIN = 0x20000000, LENGTH = 128K    /* Used for FS storage cache */
    RAM (xrw)       : ORIGIN = 0x24000000, LENGTH = 128K    /* AXI SRAM (could extend +192K from ITCM) */
    RAM_SRAM1 (xrw) : ORIGIN = 0x30000000, LENGTH = 16K     /* SRAM1 */
}
...
```

Can I just change it to the following?
```
...
/* Specify the memory areas */
MEMORY
{
    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 768K    /* sectors 0-5 */
    FLASH_APP  (rx) : ORIGIN = 0x08020000, LENGTH = 640K    /* sectors 1-5 */
    FLASH_FS (r)    : ORIGIN = 0x080c0000, LENGTH = 256K    /* sectors 6-7 */
    DTCM (xrw)      : ORIGIN = 0x20000000, LENGTH = 128K    /* Used for FS storage cache */
    RAM (xrw)       : ORIGIN = 0x24000000, LENGTH = 432K    /* AXI SRAM (could extend +192K from ITCM) */
    RAM_SRAM1 (xrw) : ORIGIN = 0x30000000, LENGTH = 16K     /* SRAM1 */
}
...
```

**Reply by jimmo** [reply]
@saraverbeecke This needs to be configured in the flash option bytes. Although I think the default value does give you 64kiB ICTM and therefore the additional 192kiB for AXI SRAM.

See the heading "RAM shared between ITCM and AXI RAM" in the reference manual (ST RM0468).

**Reply by saraverbeecke** [reply]
@jimmo thanks for looking into this. But what do you mean by "needs to be configured in the flash option bytes".
Do you mean I add it to the DTCM (xrw) line instead of the AXI SRAM?
```
..
/* Specify the memory areas */
MEMORY
{
    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 768K    /* sectors 0-5 */
    FLASH_APP  (rx) : ORIGIN = 0x08020000, LENGTH = 640K    /* sectors 1-5 */
    FLASH_FS (r)    : ORIGIN = 0x080c0000, LENGTH = 256K    /* sectors 6-7 */
    DTCM (xrw)      : ORIGIN = 0x20000000, LENGTH = 432K    /* Used for FS storage cache */
    RAM (xrw)       : ORIGIN = 0x24000000, LENGTH = 128K  /* AXI SRAM (could extend +192K from ITCM) */
    RAM_SRAM1 (xrw) : ORIGIN = 0x30000000, LENGTH = 16K     /* SRAM1 */
}
...
```

 I'm no RAM expert :-)

**Reply by jimmo** [reply, marked as answer]
Option bytes are persistent configuration registers. You can set them via st-link using e.g. Cube Programmer.

You'll need to refer to the reference manual I mentioned, but there's a two-bit setting (`TCM_AXI_SHARED`) in the option byte (`FLASH_OPTSR2_PRG`) that lets you configure how the memory layout works.

RM0468 details how this works, see page 133 for the memory layout.
0x24000000 - 0x2401FFFF (128kiB) is AXI SRAM.  (this is what the .ld file is configured to use)
0x24020000 - 0x2404FFFF (192kiB) is an additional region that may be mapped depending on the option byte setting. Anything you don't map here will be assigned to ITCM at 0x00000000.

The minimum split for ITCM (64kiB) gives the full additional 192kiB of AXI SRAM (320kiB total), the maximum split (all 192kiB for ITCM) gives none (128kiB total). This is described on page 141.

So if `TCM_AXI_SHARED` is set to 0b00, then yes you can set the length of the `RAM` region in the .ld file to 320kiB. But if it's set to 0b11 then you must leave it as it is. Like I said the default appears to be 0b00.

**Reply by jimmo** [reply]
(The important thing to note is that 0x24000000 - 0x2401FFFF is adjacent to 0x24020000 - 0x2404FFFF, so if you've mapped the additional region, then you can just treat it as one large AXI SRAM region)

**Reply by jimmo** [reply]
What MicroPython should do here is set up both regions in the linker script separately, and then when configuring the heap it should query the option byte at runtime and figure out what it is allowed to use for the heap (using the new split heap feature that allows multiple heap regions to be defined).

**Reply by saraverbeecke** [reply]
Isn't there a way the set the option bytes without a st-link (using e.g. Cube Programmer) for now, since I don't have one...

**Reply by jimmo** [reply]
The Nucleo has one built-in?

**Reply by saraverbeecke** [reply]
Hey @jimmo yep that's true I could see that the `TCM_AXI_SHARED` was set to 0b00 default. Increasing the RAM in the `.ld` file to 320K worked. I can freeze some objects in the MicroPython build so they end up in the FS storage cache part. Thank you for clearing this out. I was just wondering how to flash a custom board with the STM32H723 on it without the need for an ST-link. The settings is luckily at 0b00 so thats not needed in my case.
Anyway, can you give me any tips how I could increase the RAM even more by adding SDRAM for example. I assume I could look into the STM32F429DISC/mpconfigport.h how it is done there as the H7 also has FMC. But I'm not sure how to adapt these values from the F429DISC to the H723:
```
// SDRAM
#define MICROPY_HW_SDRAM_SIZE  (64 / 8 * 1024 * 1024)  // 64 Mbit
#define MICROPY_HW_SDRAM_STARTUP_TEST             (1)
#define MICROPY_HEAP_START  ((sdram_valid) ? sdram_start() : &_heap_start)
#define MICROPY_HEAP_END    ((sdram_valid) ? sdram_end() : &_heap_end)

// Timing configuration for 90 Mhz (11.90ns) of SD clock frequency (180Mhz/2)
#define MICROPY_HW_SDRAM_TIMING_TMRD        (2)
#define MICROPY_HW_SDRAM_TIMING_TXSR        (7)
#define MICROPY_HW_SDRAM_TIMING_TRAS        (4)
#define MICROPY_HW_SDRAM_TIMING_TRC         (7)
#define MICROPY_HW_SDRAM_TIMING_TWR         (2)
#define MICROPY_HW_SDRAM_TIMING_TRP         (2)
#define MICROPY_HW_SDRAM_TIMING_TRCD        (2)
#define MICROPY_HW_SDRAM_REFRESH_RATE       (64) // ms

#define MICROPY_HW_SDRAM_BURST_LENGTH       2
#define MICROPY_HW_SDRAM_CAS_LATENCY        3
#define MICROPY_HW_SDRAM_COLUMN_BITS_NUM    8
#define MICROPY_HW_SDRAM_ROW_BITS_NUM       12
#define MICROPY_HW_SDRAM_MEM_BUS_WIDTH      16
#define MICROPY_HW_SDRAM_INTERN_BANKS_NUM   4
#define MICROPY_HW_SDRAM_CLOCK_PERIOD       2
#define MICROPY_HW_SDRAM_RPIPE_DELAY        1
#define MICROPY_HW_SDRAM_RBURST             (0)
#define MICROPY_HW_SDRAM_WRITE_PROTECTION   (0)
#define MICROPY_HW_SDRAM_AUTOREFRESH_NUM    (4)

#define MICROPY_HW_FMC_SDCKE1   (pin_B5)
#define MICROPY_HW_FMC_SDNE1    (pin_B6)
#define MICROPY_HW_FMC_SDCLK    (pin_G8)
#define MICROPY_HW_FMC_SDNCAS   (pin_G15)
#define MICROPY_HW_FMC_SDNRAS   (pin_F11)
#define MICROPY_HW_FMC_SDNWE    (pin_C0)
#define MICROPY_HW_FMC_BA0      (pin_G4)
#define MICROPY_HW_FMC_BA1      (pin_G5)
#define MICROPY_HW_FMC_NBL0     (pin_E0)
#define MICROPY_HW_FMC_NBL1     (pin_E1)
#define MICROPY_HW_FMC_A0       (pin_F0)
#define MICROPY_HW_FMC_A1       (pin_F1)
#define MICROPY_HW_FMC_A2       (pin_F2)
#define MICROPY_HW_FMC_A3       (pin_F3)
#define MICROPY_HW_FMC_A4       (pin_F4)
#define MICROPY_HW_FMC_A5       (pin_F5)
#define MICROPY_HW_FMC_A6       (pin_F12)
#define MICROPY_HW_FMC_A7       (pin_F13)
#define MICROPY_HW_FMC_A8       (pin_F14)
#define MICROPY_HW_FMC_A9       (pin_F15)
#define MICROPY_HW_FMC_A10      (pin_G0)
#define MICROPY_HW_FMC_A11      (pin_G1)
#define MICROPY_HW_FMC_D0       (pin_D14)
#define MICROPY_HW_FMC_D1       (pin_D15)
#define MICROPY_HW_FMC_D2       (pin_D0)
#define MICROPY_HW_FMC_D3       (pin_D1)
#define MICROPY_HW_FMC_D4       (pin_E7)
#define MICROPY_HW_FMC_D5       (pin_E8)
#define MICROPY_HW_FMC_D6       (pin_E9)
#define MICROPY_HW_FMC_D7       (pin_E10)
#define MICROPY_HW_FMC_D8       (pin_E11)
#define MICROPY_HW_FMC_D9       (pin_E12)
#define MICROPY_HW_FMC_D10      (pin_E13)
#define MICROPY_HW_FMC_D11      (pin_E14)
#define MICROPY_HW_FMC_D12      (pin_E15)
#define MICROPY_HW_FMC_D13      (pin_D8)
#define MICROPY_HW_FMC_D14      (pin_D9)
#define MICROPY_HW_FMC_D15      (pin_D10)
```
